src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,185 @@
+
+namespace QGBA {
+
+LibraryEntry::LibraryEntry(mLibraryEntry* entry)
+    : entry(entry)
+    , m_fullpath(QString("%1/%2").arg(entry->base, entry->filename))
+{
+}
+
+void AbstractGameList::addEntries(QList<LibraryEntryRef> items) {
+	for (LibraryEntryRef o : items) {
+		addEntry(o);
+	}
+}
+void AbstractGameList::removeEntries(QList<LibraryEntryRef> items) {
+	for (LibraryEntryRef o : items) {
+		addEntry(o);
+	}
+}
+
+LibraryLoaderThread::LibraryLoaderThread(QObject* parent)
+	: QThread(parent)
+{
+}
+
+void LibraryLoaderThread::run() {
+	mLibraryLoadDirectory(m_library, m_directory.toUtf8().constData());
+	m_directory = QString();
+}
+
+LibraryController::LibraryController(QWidget* parent, const QString& path, ConfigController* config)
+	: QStackedWidget(parent)
+	, m_config(config)
+{
+	mLibraryListingInit(&m_listing, 0);
+
+	if (!path.isNull()) {
+		m_library = mLibraryLoad(path.toUtf8().constData());
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::startGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::startGame, this, &LibraryController::startGame);
+
+	connect(&m_loaderThread, &QThread::finished, this, &LibraryController::refresh, Qt::QueuedConnection);
+
+	refresh();
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	mLibraryListingDeinit(&m_listing);
+	if (m_library) {
+		mLibraryDestroy(m_library);
+	}
+}
+
+void LibraryController::setViewStyle(LibraryStyle newStyle) {
+	m_currentStyle = newStyle;
+
+	AbstractGameList* newCurrentWidget = nullptr;
+	if (newStyle == LibraryStyle::STYLE_LIST || newStyle == LibraryStyle::STYLE_TREE) {
+		newCurrentWidget = m_libraryTree;
+	} else {
+		newCurrentWidget = m_libraryGrid;
+	}
+	newCurrentWidget->selectEntry(selectedEntry());
+	newCurrentWidget->setViewStyle(newStyle);
+	setCurrentWidget(dynamic_cast<QWidget*>(newCurrentWidget));
+}
+
+void LibraryController::selectEntry(LibraryEntryRef entry) {
+	dynamic_cast<AbstractGameList*>(currentWidget())->selectEntry(entry);
+}
+
+LibraryEntryRef LibraryController::selectedEntry() {
+	return dynamic_cast<AbstractGameList*>(currentWidget())->selectedEntry();
+}
+
+VFile* LibraryController::selectedVFile() {
+	LibraryEntryRef entry = selectedEntry();
+	if (entry) {
+		return mLibraryOpenVFile(m_library, entry->entry);
+	} else {
+		return nullptr;
+	}
+}
+
+QPair<QString, QString> LibraryController::selectedPath() {
+	LibraryEntryRef e = selectedEntry();
+	return e ? qMakePair(e->base(), e->filename()) : qMakePair<QString, QString>("", "");
+}
+
+void LibraryController::addDirectory(const QString& dir) {
+	m_loaderThread.m_directory = dir;
+	m_loaderThread.m_library = m_library;
+	// m_loaderThread temporarily owns the library
```

Comment:
It'd be nice to have a definite on that and not an "afaik", but any reason not to have the additional safety anyway?